### PR TITLE
don't block TGB reconciliation loop on failed SG ingress reconciliation

### PR DESCRIPTION
**Pull request by michaelsaah**
### Issue

https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/3285

### Description

the controller performs an SG reconciliation step for all (cluster-wide) SGs gathered from all TGBs during the TGB reconciliation loop, before TG endpoints are reconciled: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/d177c898ddd86071eecc2fd918d72ebfb0af7892/pkg/targetgroupbinding/resource_manager.go#L139-L141

the way the code is currently written, this means that any failure during SG reconciliation blocks reconciliation of all targets across the whole cluster. such a failure can be caused by something as innocuous as a SG being deleted before the associated TGB is deleted, or a SG being entered on a TGB erroneously. this can easily lead to severe outages if not remediated quickly.

this commit changes the method `reconcileWithIngressPermissionsPerSG` to not exit on a single failed SG ingress reconciliation - instead it will log the offending error and continue through the loop.

I tested this by setting up two TGBs, modifying one to have a non-existent ingress SG, deleting a target pod, then ensuring that the new target is reconciled. the new error message was also logged:
```
{"level":"error","ts":"2023-07-26T19:39:03Z","msg":"Security group reconciliation failed for SG","SecurityGroupID":"sg-0d904125da9f258a7","error":"InvalidGroup.NotFound: The security group 'sg-0b4935e438dcca2ee' does not exist\n\tstatus code: 400, request id: 34e0001c-c85e-4f6e-b7e1-c32093afe402"}
```

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Review by johngmyers — CHANGES_REQUESTED** [review]
I cannot support what-me-worry ignoring of errors. At the very least, the error needs to be surfaced as a Kubernetes event and the handler needs to mark the work item for retry.

**Comment by michaelsaah** [comment]
thanks @johngmyers. surfacing the error via an Event sounds fine to me.

I'm not sure what you mean by "mark the work item for retry." can you point me towards any examples in the code where this is performed?

**Comment by michaelsaah** [comment]
@johngmyers @M00nF1sh just bumping this, it's an ongoing operational issue for us (Twilio) on EKS. any help appreciated.

**Comment by michaelsaah** [comment]
just another bump, any eyes appreciated

**Comment by M00nF1sh** [comment]
@michaelsaah 

Thanks for this contribution. I agree with @johngmyers that the error cannot be ignored without retry. Since such SG modification errors can be transient errors such as EC2 api throttle and shall be retried to avoid issues in security group settings.

How about below change:
1. inside reconcileWithIngressPermissionsPerSG, when reconcile for each SG(aggregatedIngressPermissionsPerSG), we don't fail upon error but aggregate all errors into a single error and return it. (e.g. use errors.Join)
2. in https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/d177c898ddd86071eecc2fd918d72ebfb0af7892/pkg/targetgroupbinding/resource_manager.go#L139, if ReconcileForPodEndpoints failed with error, we don't return the error immediately, instead we continue target registration/deregistration and return `runtime.NewRequeueNeeded("networking reconciliation")` if the ReconcileForPodEndpoints returned a error before.

**Comment by michaelsaah** [comment]
thanks for the review @M00nF1sh, addressed your feedback. have a good weekend!

**Comment by michaelsaah** [comment]
just bumping this for a re-review, thanks

**Comment by michaelsaah** [comment]
morning, just a monday review bump, thanks

File: pkg/targetgroupbinding/networking_manager.go [hunk 1]
```diff
@@ -199,11 +200,13 @@ func (m *defaultNetworkingManager) reconcileWithIngressPermissionsPerSG(ctx cont
 	aggregatedIngressPermissionsPerSG := m.computeAggregatedIngressPermissionsPerSG(ctx)
 
 	permissionSelector := labels.SelectorFromSet(labels.Set{tgbNetworkingIPPermissionLabelKey: tgbNetworkingIPPermissionLabelValue})
+	sgReconciliationErrors := make([]error, 0)
```
Comment (johngmyers) [inline pkg/targetgroupbinding/networking_manager.go:203]:
```suggestion
	var sgReconciliationErrors []error
```

**Review by johngmyers — APPROVED** [review]
Just a minor style nit

/lgtm

**Comment by michaelsaah** [comment]
thanks for the re-review @johngmyers, committed your style suggestion

**Comment by michaelsaah** [comment]
looks like that test might be flaky: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/3289#issuecomment-1646227496

**Review by M00nF1sh — APPROVED** [review]
Sorry for late review. This change looks good to me now, thanks a lot for your contribution :D 

/lgtm
/approve

**Pull request merged**